File: deno/day06.ts
import { DaySolution, parseData } from "./_tools.ts";
import { parseLines } from "./_helpers.ts";

type Grid = boolean[][];
type Position = { x: number; y: number };
const Directions = {
  U: { x: 0, y: -1 },
  R: { x: 1, y: 0 },
  D: { x: 0, y: 1 },
  L: { x: -1, y: 0 },
};
type Direction = keyof typeof Directions;

class Data {
  public grid = [] as Grid;
  public start: Position | null = null;

  public current: Position | null = null;
  public direction: Direction = "U";

  public addedObstacle: Position | null = null;
  public visited: Set<Direction>[][] = [];

  addLine(s: string) {
    this.grid.push(
      s.split("").map((c, x) => {
        if (c === "^") this.start = { x, y: this.grid.length };
        return c === "#";
      }),
    );
  }

  reset() {
    this.direction = "U";
    this.current = this.start;
    this.visited = this.grid.map((row) => row.map(() => new Set<Direction>()));
    this.visited[this.start!.y][this.start!.x].add(this.direction);
  }

  width() {
    return this.grid[0].length;
  }

  height() {
    return this.grid.length;
  }

  isObstacle({ x, y }: Position): boolean {
    return this.grid[y][x] ||
      (x === this.addedObstacle?.x && y === this.addedObstacle?.y);
  }

  isStart({ x, y }: Position): boolean {
    return x === this.start!.x && y === this.start!.y;
  }

  next(): boolean | "LOOP" {
    let { x, y } = this.current!;
    x += Directions[this.direction].x;
    y += Directions[this.direction].y;

    if (x < 0 || x >= this.width() || y < 0 || y >= this.height()) {
      return false;
    }

    if (this.visited[y][x].has(this.direction)) {
      return "LOOP";
    }

    if (this.isObstacle({ x, y })) {
      this.direction = this.turnRight();
    } else {
      this.current = { x, y };
    }
    this.visited[this.current!.y][this.current!.x].add(this.direction);

    return true;
  }

  turnRight() {
    const toRight: Record<Direction, Direction> = {
      U: "R",
      R: "D",
      D: "L",
      L: "U",
    };
    return toRight[this.direction];
  }

  walkInto(): number {
    this.reset();
    let next: boolean | "LOOP" = true;
    while (next === true) {
      next = this.next();
      if (next === "LOOP") {
        return -1;
      }
    }
    return this.visited.flat().filter((s) => s.size > 0).length;
  }
}

export class Day06 implements DaySolution<Data, number> {
  day = 6;

  parseData(dataLabel: "Example" | "Input"): Data {
    const data = new Data();
    parseLines(parseData(this.day, dataLabel), (line) => data.addLine(line));
    return data;
  }

  part1(data: Data): number {
    return data.walkInto();
  }

  part2(data: Data): number {
    let loops = 0;
    for (let y = 0; y < data.height(); y++) {
      for (let x = 0; x < data.width(); x++) {
        data.addedObstacle = null;
        if (!data.isObstacle({ x, y }) && !data.isStart({ x, y })) {
          data.addedObstacle = { x, y };
          const result = data.walkInto();
          if (result === -1) {
            loops++;
          }
        }
      }
    }
    return loops;
  }
}

File: rust/src/day06.rs
use std::collections::HashSet;
use crate::tools::read_raw_data;
use crate::DaySolution;

const OBSTACLE: char = '#';
const GUARD: [char; 4] = ['^', '>', 'v', '<'];

#[derive(Clone, Debug)]
struct Position {
    x: usize,
    y: usize,
}

#[derive(PartialEq, Debug)]
enum Op {
    TurnRight,
    Forward,
    NoOp,
}

#[derive(Clone, Debug, PartialEq, Eq, Hash)]
enum Direction {
    Up,
    Right,
    Down,
    Left,
}
impl Direction {
    fn from(c: &char) -> Direction {
        match c {
            '^' => Direction::Up,
            '>' => Direction::Right,
            'v' => Direction::Down,
            '<' => Direction::Left,
            _ => panic!("Invalid direction"),
        }
    }
}

#[derive(Clone)]
pub struct Data {
    grid: Vec<Vec<bool>>,
    current: Position,
    direction: Direction,
}

impl Data {
    fn new() -> Self {
        Data {
            grid: Vec::new(),
            current: Position { x: 0, y: 0 },
            direction: Direction::Up,
        }
    }

    fn add_line(&mut self, line: &str) {
        let row: Vec<bool> = line.chars().map(|c| { c == OBSTACLE }).collect();
        GUARD.iter().for_each(|&guard| {
            if let Some(idx) = line.find(guard) {
                self.direction = Direction::from(&guard);
                self.current = Position {
                    x: idx,
                    y: self.grid.len(),
                };
            }
        });

        self.grid.push(row);
    }

    fn width(&self) -> usize {
        self.grid[0].len()
    }

    fn height(&self) -> usize {
        self.grid.len()
    }

    fn is_obstacle(&self, x: usize, y: usize) -> bool {
        self.grid[y][x]
    }

    fn next(&mut self) -> Op {
        let Position { x, y } = self.current;
        match self.direction {
            Direction::Up => {
                if y == 0 {
                    Op::NoOp
                } else if self.is_obstacle(x, y - 1) {
                    self.direction = Direction::Right;
                    Op::TurnRight
                } else {
                    self.current.y -= 1;
                    Op::Forward
                }
            }
            Direction::Right => {
                if x == self.width() - 1 {
                    Op::NoOp
                } else if self.is_obstacle(x + 1, y) {
                    self.direction = Direction::Down;
                    Op::TurnRight
                } else {
                    self.current.x += 1;
                    Op::Forward
                }
            }
            Direction::Down => {
                if y == self.height() - 1 {
                    Op::NoOp
                } else if self.is_obstacle(x, y + 1) {
                    self.direction = Direction::Left;
                    Op::TurnRight
                } else {
                    self.current.y += 1;
                    Op::Forward
                }
            }
            Direction::Left => {
                if x == 0 {
                    Op::NoOp
                } else if self.is_obstacle(x - 1, y) {
                    self.direction = Direction::Up;
                    Op::TurnRight
                } else {
                    self.current.x -= 1;
                    Op::Forward
                }
            }
        }
    }

    fn walk_into(&self) -> Option<Vec<Vec<HashSet<Direction>>>> {
        let mut visited = vec![vec![HashSet::new(); self.width()]; self.height()];

        let mut input = self.clone();
        visited[*&input.current.y][*&input.current.x].insert(input.direction.clone());

        while Op::NoOp != input.next() {
            if visited[*&input.current.y][*&input.current.x].contains(&input.direction.clone()) {
                return None;
            }
            visited[*&input.current.y][*&input.current.x].insert(input.direction.clone());
        }

        Some(visited)

    }
}

pub struct Day06Solution;

impl DaySolution<Data, usize> for Day06Solution {
    fn day(&self) -> u16 {
        6
    }

    fn parse_input(&self, example: bool) -> Data {
        read_raw_data(self.day(), example)
            .lines()
            .fold(Data::new(), |mut data, line| {
                data.add_line(line);
                data
            })
    }

    fn part1(&self, input: &Data) -> usize {
        input.walk_into().unwrap().iter()
            .flatten()
            .filter(|directions| directions.len() >= 1)
            .count()
    }


    fn part2(&self, input: &Data) -> usize {
        input.grid.iter()
            .enumerate()
            .map(|(y, row)| row.iter().enumerate().map(move |(x, is_obstacle)| {(x, y, is_obstacle)}))
            .flatten()
            .filter(|(_, _, is_obstacle)| !**is_obstacle)
            .filter(|(x, y, _)| {
                let mut data = input.clone();
                if data.grid[*y][*x] {
                    return false;
                }
                data.grid[*y][*x] = true;
                data.walk_into().is_none()
            })
            .count()
    }
}
